src/main/kotlin/org/jitsi/nlj/util/observable/map/ObservableMap.kt

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.util.observable.map
+
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.function.BiFunction
+import java.util.function.Function
+
+interface MapEventHandler<T, U> {
+    fun entryAdded(newEntry: Map.Entry<T, U>, currentState: Map<T, U>)
+    fun entryUpdated(updatedEntry: Map.Entry<T, U>, currentState: Map<T, U>)
+    fun entryRemoved(removedEntry: Map.Entry<T, U>, currentState: Map<T, U>)
+}
+
+private class MyEntry<T, U>(
+    override val key: T,
+    override val value: U
+) : Map.Entry<T, U>
+
+/**
+ * A [MutableMap] which will notify observers (installed via [onChange]) of any changes to the map
+ * by invoking the given handler with the current state of the map.  It should be noted that some
+ * operations of this Map may be implemented in less efficient methods in order to make
+ * observability simpler.  For example, [ObservableMap.clear] is implemented by removing each
+ * entry one by one.
+ *
+ * NOTE: It was considered to invoke the handlers outside of the lock, but doing so could
+ * result in duplicate/out-of-order events being fired to the handler, so I kept them inside the lock.
+ */
+class ObservableMap<T, U>(private val data: MutableMap<T, U> = mutableMapOf()) : MutableMap<T, U> by data {
+    private val handlers: MutableList<MapEventHandler<T, U>> = CopyOnWriteArrayList()
+    /**
+     * Lock used to make all modifications to the map and notification of handlers
+     * atomic
+     */
+    private val lock = Any()
+
+    fun onChange(handler: MapEventHandler<T, U>) {
+        handlers.add(handler)
```

Comment:
I guess some changes here are missing, since `MapEventValueFilterHandler` is not here, but I wonder whether the handler gets notified of existing entries when it is first added.

Reply:
i didn't do that as part of observablemap, as it didn't seem to fit well with the idea of an observable map (which is perhaps an argument this abstraction is inappropriate here? i dunno).  i'd have to fire a series of 'adds', though the 'current state' for each one would be the entire map.  i could add it, but i dunno that the use case here requires it.  i'll think about it some more.

Reply:
Okay, I thought the previous code was explicitly handling this case. It seems that it only works because of the order things are initialized (the listeners subscribe before we receive any stream information), and that makes me wary, but it's probably fine.